cpp/autosar/src/rules/A18-0-1/CLibraryFacilitiesNotAccessedThroughCPPLibraryHeaders.ql

```diff
@@ -28,12 +28,13 @@ where
    *    not use any of 'signal.h's facilities, for example.
    */
 
-  filename = i.getIncludedFile().getBaseName() and
+  filename = i.getIncludeText().substring(1, i.getIncludeText().length() - 1) and
   filename in [
       "assert.h", "ctype.h", "errno.h", "fenv.h", "float.h", "inttypes.h", "limits.h", "locale.h",
       "math.h", "setjmp.h", "signal.h", "stdarg.h", "stddef.h", "stdint.h", "stdio.h", "stdlib.h",
       "string.h", "time.h", "uchar.h", "wchar.h", "wctype.h"
-    ]
+    ] and
+  not exists(i.getIncludedFile().getRelativePath())
```

Comment:
Excluding non-source files doesn't have a meaningful effect on the results in OpenPilot, but it would lead to FN in projects that include the C library standard headers. What are your thoughts on removing this heuristic?

Reply:
I think this really depends on what is common practice, and I dont feel like I have enough experience of what is the most common case...

If a C library standard header is checked into a project, where is the most likely location that it will be located? and should we take that into consideration?

for example the `#include "lib/assert.h"` testcase that I added, assumes that this is not a std lib, by looking at the full path in the filename
but that wouldnt work for if its flat (now added a testcase for that)

I think we actually do want to permit both of the (new) testcases because we cannot assume if these are user made or not (or at least treat them the same regardless of where they are)

Reply:
I would assume they will still use `#include <assert.h>` or `#include "assert.h"` and not `#include "lib/assert.h"` when the stdlib is part of the source tree. So the location is not important and we can keep the path heuristic.

I'm not sure if more complicated heuristics for `#include "assert.h"` are worth it. For example, sampling definitions, because you are likely required to look at transitive includes given how these standard libraries are implemented and then you have to consider whether the definition was not introduced from the cpp header file (e.g. `include <cassert>`).

To conclude, I think the path heuristic is sufficient and works for both cases where the standard library is part of the source tree and when it is not.

Reply:
asked for confirmation, but your assumptions make sense and I agree,

have reverted the one relative path heuristic and only use the name path portion